### PR TITLE
RFC refactor: use structures for `Multiplicative` and `Additive`

### PR DESCRIPTION
**Pull request by urkud**
- [x] depends on: #1214 
- [x] depends on: #1217

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#1214
* leanprover-community/mathlib4#1217
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1214~~
* ~~leanprover-community/mathlib4#1217~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This was briefly discussed on [zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/rw.20seeing.20through.20type.20synonym.3F/near/317858789).

File: Mathlib/Algebra/Order/Monoid/ToMulBot.lean [hunk 1]
```diff
@@ -56,17 +57,21 @@ theorem toMulBot_coe_ofAdd (x : α) :
...
 #align with_zero.to_mul_bot_le WithZero.toMulBot_le
 
 @[simp]
 theorem toMulBot_lt : toMulBot a < toMulBot b ↔ a < b :=
-  Iff.rfl
+  lt_iff_lt_of_le_iff_le' (toMulBot_le _ _) (toMulBot_le _ _)
 #align with_zero.to_mul_bot_lt WithZero.toMulBot_lt
 
+theorem toMulBot_mono : Monotone (@toMulBot α _) :=
+  fun _ _ => (toMulBot_le _ _).2
```
Comment (kbuzzard) [inline Mathlib/Algebra/Order/Monoid/ToMulBot.lean:71]: We're diverging from mathlib3 here, right? But probably it doesn't matter.

**Review by kbuzzard — COMMENTED** [review]
I encouraged Yury to try this refactor because now we can't mark things as irreducible after the fact, I think the one-field-structure approach for copies of types is more robust than the type synonym approach. However I now realise that perhaps `additive` and `multiplicative` are not marked `irreducible`? Are there any `local attribute [semireducible]` or `attribute [irreducible]` in mathlib3 which this PR helps with?

**Comment by semorrison** [comment]
Closed per https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/rw.20seeing.20through.20type.20synonym.3F/near/370592044

**Pull request closed without merging**